bug: startup monitor misattributes failure as 'loom MCP not connected' when a different MCP fails

## Problem

When a non-loom MCP server fails to start (e.g. a global MCP in `~/.claude.json` points to a missing binary), Claude shows "1 MCP server failed" in its UI. The startup monitor in `claude-wrapper.sh` detects this and polls `~/.claude/debug/latest` for `'MCP server "loom": Successfully connected'`.

If that message is not found within the `STARTUP_GRACE_PERIOD` (default: 20s via `LOOM_STARTUP_GRACE_PERIOD`), the monitor kills the session with:

```
Startup monitor: loom MCP not connected after 20s — killing degraded session
```

This diagnosis is **wrong** — loom may have connected fine, but the debug log check is unreliable when multiple Claude sessions run concurrently (the `latest` symlink may point to a different session). The real failing server (e.g. `sphere`) is never mentioned in the kill log line, making it hard to diagnose.

## Root Cause

The grace-period path (lines ~975–1046 of `defaults/scripts/claude-wrapper.sh`) reaches the timeout and falls through to line 1083:

```bash
log_warn "Startup monitor: loom MCP not connected after ${STARTUP_GRACE_PERIOD}s — killing degraded session"
```

This line fires even when the actual failure is a *different* MCP (e.g. `sphere`). There are two separate failure sub-cases here:

1. **Debug log not fresh**: The `~/.claude/debug/latest` symlink still points to the prior session's log (see issue #2911). The freshness guard correctly blocks reading it, so loom-connected is never set to true — but the kill message still says "loom MCP not connected", not "could not determine loom MCP status (stale debug log)".

2. **Debug log is fresh but concurrent**: With multiple shepherd sessions running in parallel, `latest` may resolve to a *different* concurrent session's log. In that session, loom connected fine — but the *failed* session's loom may or may not have connected. The wrong session is read, misattributing the failure.

In either case, the actionable information (which non-loom server failed) is already present in the debug log under patterns like:

```
MCP server "sphere": Connection failed — Cannot find module '...dist/index.js'
```

...but this is never surfaced in the kill log line.

## Impact

- 3 of 5 retry attempts burned before the root cause was identified
- Error message points at loom when a completely different MCP is broken
- Total ~10 minutes of wasted retries per shepherd invocation
- The actual broken server (`sphere`) and its error message are never logged

## Expected Behavior

The startup monitor should:

1. **Identify which MCP server(s) actually failed** by scanning the debug log (when fresh) for `MCP server "...": Connection failed` / `... failed to connect` patterns — not just the `loom` server — and include the server name(s) in the kill log line.

2. **Distinguish diagnostic confidence**: Log a different message when the debug log is stale/from another session vs. when loom is genuinely unreachable. Examples:
   - `Startup monitor: loom MCP connection unconfirmable (debug log stale/concurrent) — killing degraded session`
   - `Startup monitor: sphere MCP failed (binary not found) — loom connected OK, but sphere global MCP is broken`

3. **Surface actionable output**: Extract the first error snippet for the failing server from the debug log and include it in the warn line, e.g.:
   - `Startup monitor: sphere MCP failed — Cannot find module '.../sphere-mcp/dist/index.js'`

4. **Distinguish project vs. global MCP failures properly**: The project-MCP check at lines ~1060–1068 already does this, but only after the loom-connected inner loop resolves. When the debug log is stale, this logic is never reached. Consider also running the "which server failed" scan in the stale-log timeout path.

## Context

Observed during shepherd run for issue #3023. The `sphere` MCP in `~/.claude.json` pointed to a missing `dist/index.js`. The loom MCP itself was healthy and responding correctly to JSON-RPC.

The existing code at line ~1026 already handles the fast-path for when loom is confirmed failed:

```bash
log_warn "Startup monitor: debug log confirms loom MCP failed to connect — killing session immediately"
```

This issue is about improving the *timeout* path and the *stale-log* path to be equally informative.

See `defaults/scripts/claude-wrapper.sh` lines ~937–1090 (startup monitor logic).

## Relevant Code Locations

- **`defaults/scripts/claude-wrapper.sh`**, function `start_startup_monitor()` (~line 889)
  - MCP failure detection: lines ~936–945
  - Grace period poll loop: lines ~975–1046
  - Debug log freshness guard: lines ~1002–1018
  - Kill message (the misleading one): line ~1083
  - Project-MCP all-connected check: lines ~1060–1074

- **`STARTUP_GRACE_PERIOD`** default: 20s (line 56), configurable via `LOOM_STARTUP_GRACE_PERIOD`

## Acceptance Criteria

- [ ] When the startup monitor kills a session due to an MCP failure, the log line identifies which MCP server(s) failed (not just "loom MCP not connected")
- [ ] When the debug log is stale or belongs to another concurrent session, the kill message reflects that uncertainty rather than blaming loom
- [ ] When a non-loom MCP fails and loom itself connected successfully (but the debug log timing is ambiguous), the log message includes the name of the failing server if it can be extracted from the debug log
- [ ] No behavior change when loom genuinely fails to connect (existing fast-path at line ~1024 is unaffected)
- [ ] Wrapper tests pass (see `loom-tools/tests/` and `defaults/scripts/test_claude_wrapper.sh` if present)

## Implementation Hints

- Scan the debug log (when fresh) for all `MCP server "...": [Cc]onnection [Ff]ailed` or `[Ff]ailed to connect` lines, not just loom's. Collect the failing server name(s) into a variable.
- In the grace-period timeout branch (currently line 1083), use the collected failing server names in the log message.
- Separately track whether the stale-log guard blocked evaluation (`debug_log_was_stale=true`) and use that in the message: "loom MCP not confirmed (debug log too old)" vs "loom MCP not found in current debug log".
- The `resolved_debug_log` variable is already available within the inner loop — extracting non-loom failure lines requires only a `grep -oP` or `grep -E` call on the same file.
- Keep the existing fast-path (line ~1024) unchanged; this issue only touches the timeout/stale paths.

--- a/defaults/scripts/claude-wrapper.sh
+++ b/defaults/scripts/claude-wrapper.sh
@@ -980,6 +980,10 @@ start_startup_monitor() {
                 # loom actually failed at runtime.  See issue #2911.
                 local monitor_start_time
                 monitor_start_time=$(date +%s)
+                # Track whether the debug log was ever fresh (vs always stale/
+                # concurrent).  Used below to distinguish "loom confirmed absent"
+                # from "loom status unknown due to stale log".  See issue #3031.
+                local debug_log_ever_fresh=false
 
                 while [[ "${grace_elapsed}" -lt "${STARTUP_GRACE_PERIOD}" ]]; do
                     # Session ended on its own — nothing to kill
@@ -1018,6 +1022,7 @@ start_startup_monitor() {
                         fi
 
                         if [[ "${debug_log_is_fresh}" == "true" ]]; then
+                            debug_log_ever_fresh=true
                             # Fast-path: if loom is explicitly listed as failed in the
                             # current session's debug log, kill immediately without waiting
                             # for the full grace period.  See issue #2911.
@@ -1080,7 +1085,35 @@ start_startup_monitor() {
                         log_warn "Startup monitor: project MCP failure detected — killing session for clean restart (see issue #2652)"
                     fi
                 else
-                    log_warn "Startup monitor: loom MCP not connected after ${STARTUP_GRACE_PERIOD}s — killing degraded session"
+                    # Build informative kill message: identify which MCP(s) actually
+                    # failed rather than always blaming loom.  See issue #3031.
+                    local _mcp_fail_lines="" _failed_mcps="" _fail_detail=""
+                    if [[ -n "${resolved_debug_log:-}" ]] && [[ -f "${resolved_debug_log}" ]]; then
+                        _mcp_fail_lines=$(grep -E 'MCP server "[^"]+".*([Cc]onnection [Ff]ailed|[Ff]ailed to connect)' \
+                            "${resolved_debug_log}" 2>/dev/null || true)
+                        if [[ -n "${_mcp_fail_lines}" ]]; then
+                            _failed_mcps=$(printf '%s\n' "${_mcp_fail_lines}" | \
+                                grep -oE 'MCP server "[^"]+"' | grep -oE '"[^"]+"' | \
+                                tr -d '"' | grep -v '^loom$' | sort -u | head -3 | \
+                                tr '\n' ',' | sed 's/,$//')
+                            _fail_detail=$(printf '%s\n' "${_mcp_fail_lines}" | head -1 | \
+                                grep -oE 'Cannot find module[^;|]*|ENOENT[^;|]*|spawn ENOENT[^;|]*' | \
+                                head -1 | sed 's/[[:space:]]*$//' | cut -c1-80 || true)
+                        fi
+                    fi
+                    if [[ "${debug_log_ever_fresh}" == "false" ]]; then
+                        if [[ -n "${_failed_mcps}" ]]; then
+                            log_warn "Startup monitor: loom MCP not confirmed (debug log stale/concurrent); '${_failed_mcps}' also failed — killing degraded session"
+                        else
+                            log_warn "Startup monitor: loom MCP not confirmed after ${STARTUP_GRACE_PERIOD}s (debug log stale/concurrent) — killing degraded session"
+                        fi
+                    elif [[ -n "${_failed_mcps}" ]]; then
+                        local _detail_suffix=""
+                        [[ -n "${_fail_detail}" ]] && _detail_suffix=" — ${_fail_detail}"
+                        log_warn "Startup monitor: '${_failed_mcps}' MCP failed${_detail_suffix} — loom not confirmed — killing degraded session"
+                    else
+                        log_warn "Startup monitor: loom MCP not connected after ${STARTUP_GRACE_PERIOD}s — killing degraded session"
+                    fi
                 fi
                 pkill -INT -P $$ -f "claude" 2>/dev/null || true
                 break

